### ISSUE TITLE
[nginx-alpine] CPAD_API_DOMAIN and CPAD_FILES_DOMAIN are not applied to nginx configuration

### ISSUE THREAD

**Issue by Loader009** [xwiki-labs/cryptpad-docker#37]
Connectivity issues because of wrong domain names in nginx config.

Please correct `/etc/nginx/conf.d/cryptpad.conf`:
```
    set $api_domain "api.cryptpad...";
    set $files_domain "files.cryptpad...";
```
The following variables has been set:
```
CPAD_API_DOMAIN= "cryptpad-api.mydomain.com"
CPAD_FILES_DOMAIN="cryptpad-files.mydomain.com"
```
I use a wildcard domain, which doesn't include xxx.cryptpad.mydomain.com - thus the setting in the nginx config won't work for me.

**Comment by Loader009** [comment]
This is also the case for the `server_name` line.

**Comment by almereyda** [comment]
By the way this is defined in

https://github.com/xwiki-labs/cryptpad-docker/blob/72dd7030c1dc1c70b5ff3f53b8451f5af19a2927/docker-entrypoint.sh#L47-L70

the `CPAD_API_DOMAIN` and `CPAD_FILES_DOMAIN` *MUST* be of the subdomains `api` and `files` below `CPAD_MAIN_DOMAIN`.

Please also refer to

https://github.com/xwiki-labs/cryptpad/blob/0b78295460114c9398bd5792d7b8037e51e7079d/docs/example.nginx.conf#L10-L32

when investigating what might go wrong with your setup. Please ask, if there's something that's not obvious until here.

My suggestion would be to try to stick to the official installation instructions and examples as close as possible, and to voice further issues, when they occure.

If you leave out the ellipsis `...` from your snippet in

> Please correct /etc/nginx/conf.d/cryptpad.conf:
>
>     set $api_domain "api.cryptpad...";
>     set $files_domain "files.cryptpad...";

do those feature your `CPAD_MAIN_DOMAIN`, as expected, behind the preconfigured subdomains?

And if that is the case, why would you need to change the names of the subdomains? Could this work for you, too?

Key | Value
--- | ---
`CPAD_MAIN_DOMAIN` | `cryptpad.mydomain.com`
`CPAD_API_DOMAIN` | `api.cryptpad.mydomain.com`
`CPAD_FILES_DOMAIN` | `files.cryptpad.mydomain.com`

Eventually this might not work with a wildcard DNS RR. The production setup instructions at https://docs.cryptpad.fr/en/admin_guide/installation.html#domains also don't hint at this possibility. Yet they also don't mention `api` and `files` subdomains.

The instructions here also don't list `API_DOMAIN` and `FILES_DOMAIN` as required settings: https://github.com/xwiki-labs/cryptpad-docker/blob/72dd7030c1dc1c70b5ff3f53b8451f5af19a2927/README.md#environment-variables

Is there a specific reason for you to change them?

A more general question to the audience could be, if there are known examples on how to use these settings in a working environment.

**Comment by Loader009** [comment]
> By the way this is defined in
> 
> https://github.com/xwiki-labs/cryptpad-docker/blob/72dd7030c1dc1c70b5ff3f53b8451f5af19a2927/docker-entrypoint.sh#L64
> 
> https://github.com/xwiki-labs/cryptpad-docker/blob/72dd7030c1dc1c70b5ff3f53b8451f5af19a2927/docker-entrypoint.sh#L68
> 
> the `CPAD_API_DOMAIN` and `CPAD_FILES_DOMAIN` are set to an ambiguous value `\$\{main_domain\}`, eventually escaped into `${main_domain}`, which does not appear anywhere else in the code.
> 
>     * [ ]  This should probably be the name of these values instead.

I cannot decrypt your sed command.
`$main_domain` is set to `CPAD_MAIN_DOMAIN` as far as I understood.
Thus `CPAD_API_DOMAIN` and `CPAD_FILES_DOMAIN` have not been used here?
https://github.com/xwiki-labs/cryptpad/blob/0b78295460114c9398bd5792d7b8037e51e7079d/docs/example.nginx.conf#L19

> Please also refer to
> 
> https://github.com/xwiki-labs/cryptpad/blob/0b78295460114c9398bd5792d7b8037e51e7079d/docs/example.nginx.conf#L10-L32
> 
> when investigating what might go wrong with your setup. Please ask, if there's something that's not obvious until here.

`$main_domain` has been set to `CPAD_MAIN_DOMAIN`
https://github.com/xwiki-labs/cryptpad/blob/0b78295460114c9398bd5792d7b8037e51e7079d/docs/example.nginx.conf#L23
This line shows: __Using dedicated domains (or subdomains)__
Thus dedicated domains (not just subdomains) should be possible?

> My suggestion would be to try to stick to the official installation instructions and examples as close as possible, and to voice further issues, when they occure.
> 
> If you leave out the ellipsis `...` from your snippet in
> 
> > Please correct /etc/nginx/conf.d/cryptpad.conf:
> > ```
> > set $api_domain "api.cryptpad...";
> > set $files_domain "files.cryptpad...";
> > ```
> 
> do those feature your `CPAD_MAIN_DOMAIN`, as expected, behind the preconfigured subdomains?

The containers nginx config gets expanded to
`api.CPAD_MAIN_DOMAIN` -> `api.cryptpad.mydomain.com`
`files.CPAD_MAIN_DOMAIN` -> `files.cryptpad.mydomain.com`
Not the ones I've set via the environment variables for the container.

> And if that is the case, why would you need to change the names of the subdomains? Could this work for you, too?
> Key 	Value
> `CPAD_MAIN_DOMAIN` 	`cryptpad.mydomain.com`
> `CPAD_API_DOMAIN` 	`api.cryptpad.mydomain.com`
> `CPAD_FILES_DOMAIN` 	`files.cryptpad.mydomain.com`

No, reason is in the next answer.

> Eventually this might not work with a wildcard DNS RR. The production setup instructions at https://docs.cryptpad.fr/en/admin_guide/installation.html#domains also don't hint at this possibility. Yet they also don't mention `api` and `files` subdomains. Is this a hint at some possibility for further scaling CryptPad instances by undocumented means?
> 
> The instructions here also don't list `CPAD_API_DOMAIN` and `CPAD_FILES_DOMAIN` as required settings: https://github.com/xwiki-labs/cryptpad-docker/blob/72dd7030c1dc1c70b5ff3f53b8451f5af19a2927/README.md#environment-variables
> 
> Is there a specific reason for you to change them?

The reason for using only subdomains and not sub-subdomains:
I have a wildcard A/AAAA-entry to my server on the DNS zone -> *.mydomain.com
I try to avoid unnecessary maintenance and further entries.
A sub-subdomain would require another entry in the DNS zone -> A/AAAA for *.cryptpad.mydomain.com
I don't want this.

The same should also be the case for the SSL certificate - I use a wildcard cert to ease maintenance.
If I remember it correctly, `*.mydomain.com` does not work for `*.cryptpad.mydomain.com`.

> A more general question to the audience could be, if there are known examples on how to use these settings in a working environment.

I have one, my old server (from which I am migrating away) is having a non-docker/podman instance running.
cryptpad is running as a separate user and nginx is set up (correctly) to reverse proxy everything like it needs to.
If you want to see that config, ping me - the server will be gone on Tuesday.

Here is a piece of the config, which I already copied to show:
```
        server_name             cryptpad.mydomain.com cryptpad-api.mydomain.com cryptpad-files.mydomain.com cryptpad-ui.mydomain.com;
[…]
        set $main_domain        "cryptpad.mydomain.com";
        set $sandbox_domain     "cryptpad-ui.mydomain.com";

        set $api_domain         "cryptpad-api.mydomain.com";
        set $files_domain       "cryptpad-files.mydomain.com";
```
The config itself is a bit older, but I kept it up2date.


A little background: I am a lone admin with my own server - thus my first priority is to lower the maintenance, that's why I want to set up a docker environment - the configuration should already be set in the image/containers - and less configuration should be done outside.

**Comment by promasu** [comment]
I just cite the readme:

---
Docker images and their supporting configuration files are provided as a community effort by those using them, with support provided by the core development team on a best-effort basis. Keep in mind that the core team neither uses nor tests Docker images, so your results may vary.
---

I saw this issue but I currently have no time to look into it. I will in a couple of days.
(And I actually don't really see a reason to mention us ^^)

**Comment by Loader009** [comment]
I will also try different things out, some modified/external nginx config to play with.

But it will take at least a week, I am at the last phase of my server migration.

**Comment by almereyda** [comment]
Thank you for repeating the README.md section. It felt only natural to ping you two for the progress we made here, but I understand that we are discussing features that are not supported by you. Eventually we can rewrite some things in the README to make that fact more apparent.

When skimming through the document for technical details, the third paragraph is not the most obvious. I'll PR a little update to the README that adds a Disclaimer headline and a Waiver of warranties, so we're more safe on this side in the long run.

---

And thanks for checking the details. I'm interested in stabilising and harmonising this image (and instructions) as much as possible, why the feedback is very valuable.

**Comment by Loader009** [comment]
So - I've checked the nginx config and compared it to my old config (I did a backup ofc).
The difference is actually those things I mentioned (the domains/variables for them) and the real_ip module (which tried to avoid up untill now).

I have to test more (especially the reverse proxy thing and the forwarded IPs).

I'll try it without the real_ip module - but (keep this in mind) I am not totally aware of all security risks posing from proxy headers.
Basically I want to mostly avoid anything related to proxy headers.

I have to read into this too -> https://httpoxy.org/

> _**Immediate Mitigation**_
> 
> _The best immediate mitigation is to block Proxy request headers as early as possible, and before they hit your application. This is easy and safe._
